crates/bevy_ecs/src/template.rs

```diff
@@ -0,0 +1,190 @@
+//! Functionality that relates to the [`Template`] trait.
+
+pub use bevy_ecs_macros::GetTemplate;
+
+use crate::{
+    bundle::Bundle,
+    entity::{Entity, EntityPath},
+    error::{BevyError, Result},
+    world::EntityWorldMut,
+};
+use alloc::{boxed::Box, vec, vec::Vec};
+use bevy_platform::collections::hash_map::Entry;
+use bevy_utils::TypeIdMap;
+use core::any::{Any, TypeId};
+use downcast_rs::{impl_downcast, Downcast};
+use variadics_please::all_tuples;
+
+/// A [`Template`] is something that, given a spawn context (target [`Entity`], [`World`](crate::world::World), etc), can produce a [`Template::Output`].
+pub trait Template {
```

Comment:
Can we currently spawn resources via templates?

If no, are you planning to allow that? Is that blocked on resources-as-components?

Reply:
We could have a ResourceTemplate that has no effect on the entity in question and just inserts the Resource (and handles conflicts in whatever way it sees fit). I think this _might_ be better suited to the BSN Set scenario, where we could make resources their own "type" in the set. But theres nothing stopping us from experimenting with the ResourceTemplate approach.